pyccel/stdlib/cwrapper/cwrapper.c

```diff
@@ -0,0 +1,508 @@
+#include "cwrapper.h"
+
+
+
+// strings order needs to be the same as its equivalent numpy macro
+// https://numpy.org/doc/stable/reference/c-api/dtype.html
+const char* dataTypes[17] = {"Bool", "Int8", "UInt8", "Int16", "UIn16", "Int32", "UInt32",
+                             "Int64", "UInt64", "Int128", "UInt128", "Float32", "Float64",
+                             "Float128", "Complex64", "Complex128", "Complex256"};
+
+
+
+/*
+ * Functions : Cast functions
+ * --------------------------
+ * All functions listed down are based on C/python api
+ * with more tolerance to different precision
+ * Convert python type object to the desired C type
+ * Parameters :
+ *     object : the python object
+ * Returns    :
+ *     The desired C type, an error may be raised by c/python converter
+ *     so one should call PyErr_Occurred() to check for errors after the
+ *	   calling a cast function
+ * Reference of the used c python api function
+ * --------------------------------------------
+ * https://docs.python.org/3/c-api/float.html#c.PyFloat_AsDouble
+ * https://docs.python.org/3/c-api/complex.html#c.PyComplex_RealAsDouble
+ * https://docs.python.org/3/c-api/complex.html#c.PyComplex_ImagAsDouble
+ * https://docs.python.org/3/c-api/long.html#c.PyLong_AsLong
+ * https://docs.python.org/3/c-api/long.html#c.PyLong_AsLongLong
+ */
+
+float complex PyComplex_to_Complex64(PyObject *object)
+{
+	float complex	c;
+
+	// https://numpy.org/doc/1.17/reference/c-api.array.html#c.PyArray_IsScalar
+	// https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_ScalarAsCtype
+	if (PyArray_IsScalar(object, Complex64))
+		PyArray_ScalarAsCtype(object, &c);
+
+	else
+	{
+		float real_part = (float)PyComplex_RealAsDouble(object);
+		float imag_part = (float)PyComplex_ImagAsDouble(object);
+
+		c = real_part + imag_part * _Complex_I;
+	}
+	return	c;
+}
+//-----------------------------------------------------//
+double complex	PyComplex_to_Complex128(PyObject *object)
+{
+	double complex	c;
+	double	real_part;
+	double	imag_part;
+
+	real_part = PyComplex_RealAsDouble(object);
+	imag_part = PyComplex_ImagAsDouble(object);
+
+	c = real_part + imag_part * _Complex_I;
+
+	return	c;
+}
+//-----------------------------------------------------//
+int64_t	PyInt64_to_Int64(PyObject *object)
+{
+	int64_t		i;
+	long long	out;
+
+	out = PyLong_AsLongLong(object);
+
+	i = (int64_t)out;
+
+	return	i;
```

Comment:
why we need a `long long` intermediate variable `out`?
can't we return directly `return *(int64_t)PyLong_AsLongLong(object);`

Reply:
those functions are very small, I suggest inline them for performance, and add them the header directly,

Reply:
I agree about reducing this to one line. I knew there was something I wanted to simplify in this file but I forgot what. I think this was it :+1:

Reply:
> those functions are very small, I suggest inline them for performance, and add them the header directly,

This depends on where we want to see better performance. The cwrapper shouldn't be the bottleneck of a translated code. Inlining can be a costly operation for the compiler. So I am concerned that we may see a slowdown in the runtime of our tests and very little speedup in return. I should test this

Reply:
Fixed 7d3d8d3
Inlining to test

Reply:
Regarding inlining:
https://softwareengineering.stackexchange.com/a/56230

Reply:
It seems the runner is running slowly today so I'm not certain about the performance, but there doesn't seem to be a slowdown between [this run](https://github.com/pyccel/pyccel/actions/runs/1122914121) and [this run](https://github.com/pyccel/pyccel/actions/runs/1122931679) so I am keeping the inlining

Reply:
Inlining fixed : 812d1eb

Reply:
I added more inlining, we will see how this affects the compile time moving forward

Reply:
great.